src/main/scala/io/keen/client/scala/BatchWriterClient.scala

```diff
@@ -0,0 +1,208 @@
+ * queue is drained.
+ *
+ * Events are queued for batch submission with [[queueEvent]]; other operations
+ * like [[addEvent]] function as with an ordinary `Client with Writer`—that is,
+ * they are non-blocking but effect discrete API calls per invocation.
+ *
+ * Batch size, flush scheduling, queue bounds, etc. can be tuned via the settings
+ * under the `keen.queue` property tree.
+ *
+ * TODO: explain the difference in behavior if send-interval is zero seconds.
+ *
+ * @param config Client configuration, by default loaded from `application.conf`.
+ */
+class BatchWriterClient(config: Config = ConfigFactory.load())
+    extends Client(config)
+    with Writer {
+
+  import BatchWriterClient._
+
+  /** The number of events sent in a single API call when flushing batches */
+  val batchSize: Integer = settings.batchSize
+
+  /** Timeout for each bulk write API call when flushing batches */
+  val batchTimeout: FiniteDuration = settings.batchTimeout
+
+  /** Threshold of queued events at which flush of batches is triggered */
+  val sendIntervalEvents: Integer = settings.sendIntervalEvents
+
+  /** Time interval at which batches of queued event writes are scheduled to be flushed */
+  val sendInterval: FiniteDuration = settings.sendIntervalDuration
+
+  /** Duration for which client will wait for scheduled batch flushes to complete when shutting down */
+  val shutdownDelay: FiniteDuration = settings.shutdownDelay
+
+  // initialize and configure our local event store queue
+  // FIXME: should be protected, but tests need to be updated; the EventStore
+  // should be an injectable component like HttpAdapter (though the injection
+  // pattern still needs work on that too)
+  val eventStore: EventStore = new RamEventStore
+  eventStore.maxEventsPerCollection = settings.maxEventsPerCollection
+
+  // Schedule sending of queued events.
+  protected val scheduledThreadPool: Option[ScheduledThreadPoolExecutor] = scheduleSendQueuedEvents()
+
+  /**
+   * Queue an event for batched publishing.
+   *
+   * @param collection The collection to which the event will be added.
+   * @param event The event
+   */
+  def queueEvent(collection: String, event: String): Unit = {
+    // bypass min/max intervals for testing
+    // FIXME: There are less kludgey ways to achieve testability here
+    environment match {
+      case Some("test") if Some("test").get matches "(?i)test" =>
+      case _ =>
+        require(
+          sendIntervalEvents == 0 || (sendIntervalEvents >= MinSendIntervalEvents && sendIntervalEvents <= MaxSendIntervalEvents),
+          s"Send events interval must be between $MinSendIntervalEvents and $MaxSendIntervalEvents"
+        )
+    }
+
+    eventStore.store(projectId, collection, event)
+
+    // If we've met a configured event count threshold, flush the queue.
+    if (sendIntervalEvents != 0 && eventStore.size >= sendIntervalEvents) {
+      sendQueuedEvents()
```

Comment:
Good catch. Definitely better to use `sendQueuedEventsAsync` here.

Flushing in `queueEvent` is something specific to this implementation, although it would be nice to have it available in the Java client as well.